### PR TITLE
Add CAP html table to AAD report

### LINKED ISSUE #64: Explore integration of Conditional Access Policies into AAD HTML report
I also think we should explore dumping the tenant's CAPs into the AAD report. This would be require some effort outside the scope of this issue. It's need updates on the provider side and the report side. We'd need to set limits on numbers, etc but I think it would be really helpful.

One thing I see a lot is people complaining about how hard it is to visualize/understand a large number of CAPs.

_Originally posted by @ethanb-cisa

### PR DESCRIPTION
**Pull request by adhilto**
## 🗣 Description ##

Add a table to the end of the AAD report that allows users to quickly understand their conditional access policies (CAPs).

## 💭 Motivation and context ##
Closes #64.

## 🧪 Testing ##

1. Created ~20 different conditional access policies on Test Tenant 4 (E5), attempting to explore the full feature set. Ensured that each CAP was displayed properly.
2. Tested against Test Tenant 1 (G5) and Test Tenant 2 (G3).
3. Tested as user with no admin privileges.
4. Tested against a tenant with no CAPs (Test Tenant 4, unfortunately that means I deleted the aforementioned CAPs)
5. Tested with Pester. Created test cases that also explore the full feature set discovered in step 1.

Note that the Pester cases added here won't be run automatically yet. That will come once Powershell-Unit-Testing-85 is merged in. For now, to manually run these tests, navigate to `Testing\Unit\Pester\Providers\AADProvider`, then run `Invoke-Pester -Output detailed .\Export-Caps.Tests.ps1`.

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [X] All relevant type-of-change labels have been added.
- [X] ~~I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.~~
- [X] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [X] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [X] Tests have been added and/or modified to cover the changes in this PR.
- [X] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Add a tag or create a release.

### CODE REVIEW

**Comment by ssatyapal123** [comment]
Successfully ran the branch on the dissent G5 tenant and the AAD report looks good and is most helpful!

However, I got the following error when running the unit test per the command: Invoke-Pester -Output detailed .\Export-Caps.Tests.ps1. Am I missing a step? Thank you.

Invoke-Pester : Parameter cannot be processed because the parameter name 'Output' is ambiguous. Possible matches include: -OutputXml -OutputFile -OutputFormat.
At line:1 char:15
+ Invoke-Pester -Output detailed .\Export-Caps.Tests.ps1
+               ~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Invoke-Pester], ParameterBindingException
    + FullyQualifiedErrorId : AmbiguousParameter,Invoke-Pester

image

**Comment by adhilto** [comment]
> Successfully ran the branch on the dissent G5 tenant and the AAD report looks good and is most helpful!
> 
> However, I got the following error when running the unit test per the command: Invoke-Pester -Output detailed .\Export-Caps.Tests.ps1. Am I missing a step? Thank you.
> 
> Invoke-Pester : Parameter cannot be processed because the parameter name 'Output' is ambiguous. Possible matches include: -OutputXml -OutputFile -OutputFormat. At line:1 char:15
> 
> * Invoke-Pester -Output detailed .\Export-Caps.Tests.ps1
> * ```
>             ~~~~~~~
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   
>   * CategoryInfo          : InvalidArgument: (:) [Invoke-Pester], ParameterBindingException
>   * FullyQualifiedErrorId : AmbiguousParameter,Invoke-Pester
> 
> image

Have you run anything with Pester yet? See https://pester.dev/docs/introduction/installation for installation instructions.

**Comment by ssatyapal123** [comment]
Thanks, that helped! I had to remove the existing version (as administrator) and install the newer version of Pester. Below is the output for the cisaent. Sorry if it's obvious but how do I know if this result is what it should be? Is the purpose to assess performance (hence the times)? Thank you.

PS C:\users\ssatyapal\SCuBAGear\ScubaGear-64-cap-html-report\Testing\Unit\Pester\Providers\AADProvider> Invoke-Pester -Output detailed .\Export-Caps.Tests.ps1 > output.txt
Pester v5.4.0

Starting discovery in 1 files.
Discovery found 70 tests in 99ms.
Running tests.

Running tests from 'C:\users\ssatyapal\SCuBAGear\ScubaGear-64-cap-html-report\Testing\Unit\Pester\Providers\AADProvider\Export-Caps.Tests.ps1'
Describing GetIncludedUsers
  [+] returns 'None' when no users are included 8ms (4ms|3ms)
  [+] handles including single users 5ms (4ms|1ms)
  [+] handles including multiple users 5ms (4ms|1ms)
  [+] handles including single groups 6ms (5ms|1ms)
  [+] handles including multiple groups 6ms (4ms|2ms)
  [+] handles including single roles 5ms (4ms|1ms)
  [+] handles including multiple roles 6ms (5ms|1ms)
  [+] handles including users, groups, and roles simultaneously 5ms (4ms|1ms)
  [+] returns 'All' when all users are included 9ms (4ms|4ms)
  [+] handles including single type of external user 6ms (5ms|1ms)
  [+] handles including all types of guest users 6ms (5ms|1ms)
  [+] handles empty input 4ms (3ms|2ms)

Describing GetExcludedUsers
  [+] returns 'None' when no users are included 9ms (4ms|5ms)
  [+] handles excluding single users 5ms (4ms|1ms)
  [+] handles excluding multiple users 5ms (4ms|1ms)
  [+] handles excluding single groups 5ms (4ms|2ms)
  [+] handles excluding multiple groups 5ms (4ms|1ms)
  [+] handles excluding single roles 8ms (4ms|4ms)
  [+] handles excluding multiple roles 7ms (6ms|1ms)
  [+] handles excluding users, groups, and roles simultaneously 5ms (4ms|1ms)
  [+] handles excluding all types of external users 5ms (4ms|1ms)
  [+] handles excluding a single type of external user 5ms (4ms|1ms)
  [+] handles empty input 3ms (2ms|1ms)

Describing GetApplications
  [+] handles including all apps 10ms (6ms|4ms)
  [+] handles including/excluding no apps 7ms (6ms|1ms)
  [+] handles including/excluding single specific apps 11ms (10ms|1ms)
  [+] handles including/excluding multiple specific apps 8ms (6ms|2ms)
  [+] handles app filter in include mode 9ms (7ms|1ms)
  [+] handles app filter in exclude mode 8ms (7ms|1ms)
  [+] handles including app filter and specific apps 8ms (7ms|1ms)
  [+] handles excluding app filter and specific apps 8ms (7ms|1ms)
  [+] handles registering a device 8ms (5ms|3ms)
  [+] handles registering security info 6ms (5ms|1ms)
  [+] handles registering security info 8ms (7ms|1ms)
  [+] handles empty input 3ms (2ms|1ms)

Describing GetConditions
  [+] handles user risk levels 8ms (4ms|4ms)
  [+] handles sign-in risk levels 7ms (6ms|1ms)
  [+] handles including all device platforms 6ms (5ms|1ms)
  [+] handles including/excluding specific device platforms 6ms (5ms|1ms)
  [+] handles including all locations 8ms (7ms|1ms)
  [+] handles excluding trusted locations 6ms (5ms|1ms)
  [+] handles including/excluding single custom locations 10ms (9ms|1ms)
  [+] handles including/excluding multiple custom locations 6ms (5ms|1ms)
  [+] handles including trusted locations 7ms (6ms|1ms)
  [+] handles including all client apps 5ms (4ms|1ms)
  [+] handles including specific client apps 5ms (4ms|1ms)
  [+] handles custom client app filter in include mode 5ms (4ms|1ms)
  [+] handles custom client app filter in exclude mode 5ms (4ms|1ms)
  [+] handles many conditions simultaneously 15ms (14ms|1ms)
  [+] handles empty input 7ms (6ms|1ms)

Describing GetAccessControls
  [+] handles blocking access 8ms (4ms|4ms)
  [+] handles requiring single control 6ms (5ms|1ms)
  [+] handles requiring multiple controls in AND mode 5ms (4ms|1ms)
  [+] handles requiring multiple controls in OR mode 6ms (4ms|2ms)
  [+] handles using authentication strength (phishing resistant MFA) 4ms (4ms|1ms)
  [+] handles using both authentication strength and a traditional control 4ms (4ms|1ms)
  [+] handles using no access controls 5ms (3ms|1ms)
  [+] handles empty input 3ms (2ms|1ms)

Describing GetSessionControls
  [+] handles using no session controls 12ms (8ms|4ms)
  [+] handles using app enforced restrictions 6ms (5ms|1ms)
  [+] handles using conditional access app control with custom policy 6ms (6ms|1ms)
  [+] handles using conditional access app control in monitor mode 5ms (4ms|1ms)
  [+] handles using conditional access app control in block mode 5ms (4ms|1ms)
  [+] handles using sign-in frequency every time 5ms (4ms|1ms)
  [+] handles using sign-in frequency time based 5ms (4ms|1ms)
  [+] handles using persistent browser session 7ms (6ms|1ms)
  [+] handles using customized continuous access evaluation 5ms (4ms|1ms)
  [+] handles disabling resilience defaults 6ms (5ms|1ms)
  [+] handles multiple controls simultaneously 9ms (8ms|1ms)
  [+] handles empty input 3ms (2ms|1ms)
Tests completed in 953ms
Tests Passed: 70, Failed: 0, Skipped: 0 NotRun: 0

**Comment by adhilto** [comment]
@ssatyapal123 The important bit is the final line:

> Tests Passed: 70, Failed: 0, Skipped: 0 NotRun: 0

All the tests passed. The times don't matter that much.

**Review by ssatyapal123 — APPROVED** [review]
Looks good.

File: PowerShell/ScubaGear/Modules/Providers/ProviderHelpers/AADConditionalAccessHelper.psm1 [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+            "IncludeAuthenticationContextClassReferences", "IncludeUserActions"))
+        if ($Missing.Length -gt 0) {
+            Write-Warning "Conditional access policy structure not as expected. The following keys are missing: $($Missing -Join ', ')"
+            return @()
+        }
+
+        # Begin processing the CAP
+        $Actions = @()
+        if ($Cap.Conditions.Applications.IncludeApplications.Length -gt 0 -or
+            $null -ne $Cap.Conditions.Applications.ApplicationFilter.Mode) {
```
[the remark this answers is not in the document]
Comment (adhilto) [reply PowerShell/ScubaGear/Modules/Providers/ProviderHelpers/AADConditionalAccessHelper.psm1]: I thought I had read somewhere that the $null should be first, but I couldn't find it today, so I went ahead and applied your suggestion, as I agree it's more readable. Alas, after committing the change, I found where I read that:
image

I'll switch it back but add a comment somewhere than $null should be first to keep the linter happy.

**Comment by adhilto** [comment]
Documenting other changes David requested in a DM:

- [ ] Make the "Expand all" and "Collapse all" buttons into actual buttons
- [ ] Spice up the color a bit
- [ ] Move the SVG arrows to the left of the CAP names
- [ ] Show actual user/group/role names instead of just saying "x specific users"

**Comment by nanda-katikaneni** [comment]
Ran aad tests with the branch on couple tenants and compared outputs against main branch - except for the new conditional access policies, results are same. merging into main.

**Pull request merged**